pkg/objstore/s3/s3.go

```diff
@@ -0,0 +1,148 @@
+		bucket: *conf.Bucket,
+		client: client,
+		opsTotal: prometheus.NewCounterVec(prometheus.CounterOpts{
+			Name:        "thanos_objstore_s3_bucket_operations_total",
+			Help:        "Total number of operations that were executed against an s3 bucket.",
+			ConstLabels: prometheus.Labels{"bucket": *conf.Bucket},
+		}, []string{"operation"}),
+	}
+	if reg != nil {
+		reg.MustRegister(bkt.opsTotal)
+	}
+	return bkt, nil
+}
+
+// Iter calls f for each entry in the given directory. The argument to f is the full
+// object name including the prefix of the inspected directory.
+func (b *Bucket) Iter(ctx context.Context, dir string, f func(string) error) error {
+	b.opsTotal.WithLabelValues(opObjectsList).Inc()
+	// Ensure the object name actually ends with a dir suffix. Otherwise we'll just iterate the
+	// object itself as one prefix item.
+	if dir != "" {
+		dir = strings.TrimSuffix(dir, DirDelim) + DirDelim
+	}
+
+	// create a channel to control ListObjects goroutine
+	doneCh := make(chan struct{})
+	defer close(doneCh)
+
+	for object := range b.client.Client.ListObjects(b.bucket, dir, false, doneCh) {
+		// this sometimes happens with empty buckets
+		if object.Key == "" {
+			continue
+		}
+		if err := f(object.Key); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+// Get returns a reader for the given object name.
+func (b *Bucket) Get(ctx context.Context, name string) (io.ReadCloser, error) {
+	b.opsTotal.WithLabelValues(opObjectGet).Inc()
+	rc, _, err := b.client.GetObject(b.bucket, name, minio.GetObjectOptions{})
+	return rc, err
+}
+
+// GetRange returns a new range reader for the given object name and range.
+func (b *Bucket) GetRange(ctx context.Context, name string, off, length int64) (io.ReadCloser, error) {
+	b.opsTotal.WithLabelValues(opObjectGet).Inc()
+	opts := &minio.GetObjectOptions{}
+	err := opts.SetRange(off, off+length)
+	if err != nil {
+		return nil, err
+	}
+	rc, _, err := b.client.GetObject(b.bucket, name, *opts)
+	if err != nil {
+	}
+	return rc, err
+
+}
+
+// Exists checks if the given object exists.
+func (b *Bucket) Exists(ctx context.Context, name string) (bool, error) {
+	b.opsTotal.WithLabelValues(opObjectStat).Inc()
+	_, err := b.client.StatObject(b.bucket, name, minio.StatObjectOptions{})
+	if err != nil {
+		errResponse := minio.ToErrorResponse(err)
+		if errResponse.Code == "NoSuchKey" {
+			return false, nil
+		}
+		return false, err
+	}
+
+	return true, nil
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	b.opsTotal.WithLabelValues(opObjectInsert).Inc()
+	_, err := b.client.PutObjectWithContext(ctx, b.bucket, name, r, -1, minio.PutObjectOptions{})
+	return err
+}
+
+// Delete removes the object with the given name.
+func (b *Bucket) Delete(ctx context.Context, name string) error {
+	b.opsTotal.WithLabelValues(opObjectDelete).Inc()
+	return b.client.RemoveObject(b.bucket, name)
+}
```

Comment:
None of these operations support context-based cancelation. Looking at the minio code, it ironically calls unexported methods under-the-hood that have a context argument and passes them a background context. Are you aware whether they have plans to change/extend the unexported API to support custom contexts?

@gouthamve

Reply:
Actually looks like at least Put and Get have context variants: https://godoc.org/github.com/minio/minio-go#Client.GetObjectWithContext

Reply:
Yeah I think I was using those at some point and then changed for some reason, I'll utilize those. I can look into adding a `RemoveObjectWithContext`